src/core/gui/PagePreviewDecoration.cpp

```diff
@@ -0,0 +1,117 @@
+#include "PagePreviewDecoration.h"
+
+#include <math.h>
+
+#include "control/Control.h"
+#include "control/settings/Settings.h"
+#include "util/raii/CairoWrappers.h"
+
+void PagePreviewDecoration::drawDecoration(cairo_t* cr, SidebarPreviewPageEntry* pageEntry, Control* control) {
+    switch (control->getSettings()->getSidebarNumberingStyle()) {
+        case SidebarNumberingStyle::STYLE_1:
+            PagePreviewDecoration::drawPageNumberBelowPreview(cr, pageEntry, control);
+            break;
+        case SidebarNumberingStyle::STYLE_2:
+            PagePreviewDecoration::drawPageNumberWithCircleBackground(cr, pageEntry, control);
+            break;
+        case SidebarNumberingStyle::STYLE_3:
+            PagePreviewDecoration::drawPageNumberWithSquareBackground(cr, pageEntry, control);
+            break;
+        default:
+            break;
+    }
+}
+
+void PagePreviewDecoration::drawPageNumberBelowPreview(cairo_t* cr, SidebarPreviewPageEntry* pageEntry,
+                                                       Control* control) {
+    xoj::util::CairoSaveGuard saveGuard(cr);
+    cairo_text_extents_t extents;
+    std::string pageNumber = std::to_string(pageEntry->getIndex() + 1);
+    Color color = control->getSettings()->getBackgroundColor();
+    if (pageEntry->isSelected()) {
+        color = control->getSettings()->getBorderColor();
+    }
+    Util::cairo_set_source_rgbi(cr, color);
+    cairo_select_font_face(cr, "Sans", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
+    cairo_set_font_size(cr, FONT_SIZE);
+    cairo_text_extents(cr, pageNumber.c_str(), &extents);
+    double x = pageEntry->getWidth() / 2 - (extents.width / 2 + extents.x_bearing);
+    double y = pageEntry->getHeight() - MARGIN_BOTTOM - (extents.height / 2 + extents.y_bearing);
+    cairo_move_to(cr, x, y);
+    cairo_show_text(cr, pageNumber.c_str());
+}
+
+void PagePreviewDecoration::drawPageNumberWithCircleBackground(cairo_t* cr, SidebarPreviewPageEntry* pageEntry,
+                                                               Control* control) {
+    xoj::util::CairoSaveGuard saveGuard(cr);
+    cairo_text_extents_t extents;
+    std::string pageNumber = std::to_string(pageEntry->getIndex() + 1);
+    cairo_select_font_face(cr, "Sans", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
+    cairo_set_font_size(cr, FONT_SIZE);
+    cairo_text_extents(cr, pageNumber.c_str(), &extents);
+
+    // Set the color for drawing the circle. This circle is required to provide good contrast for the page number.
+    if (pageEntry->isSelected()) {
+        Color color = control->getSettings()->getBorderColor();
+        Util::cairo_set_source_rgbi(cr, color, 0.6);
+    } else {
+        cairo_set_source_rgba(cr, 0.3, 0.3, 0.3, 0.6);
```

Comment:
and colors as well
```suggestion
        Util::cairo_set_source_rgbi(cr, Colors::xopp_darkgrey, LABEL_OPACITY);
```
and Colors::xopp_darkgrey is to be defined in util/include/util/Color.h (darkgrey might not be the best name, I don't know)

Reply:
> and colors as well
> 
> and Colors::xopp_darkgrey is to be defined in util/include/util/Color.h (darkgrey might not be the best name, I don't know)

I would simply take the already defined ` Color xopp_darkslategray{0x434343U};`
It corresponds to (0.263,0.263,0.263) and is therefore a tad darker than (0.3,0.3,0.3), but not that much. 
We actually have a script (color-names) in the `scripts` folder to determine color names. In this case (running it via `python3 color-names 4D4D4D`) it returned `darkslategray` as the closest approximation.